Major: re-wrote code descriptions (#16500)
mdn/content: files/en-us/web/javascript/reference/global_objects/webassembly/table/table/index.md

### Creating a new WebAssembly Table instance

The following example creates a new WebAssembly Table instance with an initial size of 2

> elements. The WebAssembly Table is created and accessed in JavaScript while also visible and callable inside a wasm instance.

Reviewer: Makes sense. I've updated the suggestion.
Nitpick: As per https://webassembly.org/ , the abbreviation has capital "W" in it. So, unless it's a file extension we need to use "Wasm" everywhere. thoughts?

> elements. The WebAssembly Table is created and accessed in JavaScript while also visible and callable inside a wasm instance.

Reviewer 2: Yeah. I wasn't sure about the capitalization, but it seems like "Wasm" is better. I prefer "WebAssembly" just because "Wasm" is so ugly. But I can appreciate that that's a personal thing :).

This example uses the following reference files: 
1. `table2.html`:  An HTML file containing a script that loads and instantiates an external [`WebAssembly.Table()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Table/Table) ([source code](https://github.com/mdn/webassembly-examples/blob/master/js-api-examples/table2.html) and [live version](https://mdn.github.io/webassembly-examples/js-api-examples/table2.html))
2. `table2.wasm`: The Wasm module being imported in table2.html ([source code](https://github.com/mdn/webassembly-examples/blob/master/js-api-examples/table2.wat))

> In `table2.html`, a `WebAssembly.Table()` is instantiated with the following: 

Reviewer: > we create a `Table` is instantiated with the following code

Doesn't sound right.

How about this?
```suggestion
In `table2.html`, we create a `Table` using following code: 
```

> In `table2.html`, a `WebAssembly.Table()` is instantiated with the following: 

Reviewer 2: Almost! But you need a "the" in between "using" and "following" :).

```js
const tbl = new WebAssembly.Table({
  initial: 2, 
  element: "anyfunc"
});
console.log(tbl.length);  // a table with 2 elements
console.log(tbl.get(0));  // content for index 1 is null
console.log(tbl.get(1));  // content for index 2 is null
```

The index contents are retrieved using {{jsxref("WebAssembly/Table/get", "Table.prototype.get()")}}.

Create an import object that contains the table:

```js
const importObject = {
  js: {
    tbl:tbl
  }
};
```

Load and instantiate a wasm module using the {{jsxref("WebAssembly.instantiateStreaming()")}} method:  

```js
WebAssembly.instantiateStreaming(fetch('table2.wasm'), importObject)
.then(function(obj) {
  console.log(tbl.length);  // table length is still 2
  console.log(tbl.get(0)());  // content for index 1 is 42 and is an Exported WebAssembly function; note the additional '()'
  console.log(tbl.get(1)());  // content for index 2 is 83 and is an Exported WebAssembly function; note the additional '()'
});

> ```

Reviewer 2: Again this is tricky because we're doing two things here:
- instantiating the module
- logging the results

It would be very nice to be able to talk about what the WebAssembly module does before the console.log statements. So perhaps we should split this code block too?

````
Next, we load and instantiate a WebAssembly module using the {{jsxref("WebAssembly.instantiateStreaming()")}} method:

```js
const instantiating = WebAssembly.instantiateStreaming(fetch('table2.wasm'), importObject);
```

The `table2.wasm` module contains two functions, one returning 42 and one returning 83. It stores the functions in elements 0 and 1 of the imported table.

Once the module is instantiated we can call the functions in the table:

```js
instantiating.then( obj => {
  console.log(tbl.length);
  console.log(tbl.get(0)());    // 42
  console.log(tbl.get(1)());    // 83
});
```

````
?

The `table2.wasm` module contains two functions:
1. Return 42 and store this value in index 1 of the `importObject` table
2. Return 83 and store value in index 2 of `importObject` table

After instantiating `table2.wasm`, `tbl` is updated with the following:
- table length is still 2
- content for index 1 is now a function which returns 42
- content for index 2 is now a function which returns 83

> The content for indexes 1 and 2 are now callable [Exported WebAssembly Functions](/en-US/docs/WebAssembly/Exported_functions), which are called to change the index's content values. To call the wasm functions directly from `tbl`, append a second function invocation operator at the end of the

Reviewer 2: I don't think the functions "change the index's content values" - they just return the values to the caller. So when you do `console.log(tbl.get(0)());` you're calling the function at index 0 and logging the thing it returned. 

As far as I can tell this is not really any different from anywhere else you might store a function, for example:

```
function hello() {
   return "hello!";
}

const array = [hello];

console.log(array[0]()); // "hello!"
```

....and it's just equivalent to:

```
function hello() {
   return "hello!";
}

const array = [hello];
const helloFunction = array[0];

console.log(helloFunction()); // "hello!"
```

In just the same way, we could replace this example code:

```js
console.log(tbl.get(0)());
```

...with this:

```js
const fn1 = tbl.get(0);
console.log(fn1());
```

So I think if we want to make a point of the caller adding `()`, we might just say something like:

```
The items at indexes 0 and 1 of the table are functions, so to call them, note that we must add the function invocation operator after the `get()` call.
```
